Sources/Sentry/SentryProfiler.mm

```diff
@@ -421,12 +293,150 @@
     return payload;
 }
 
+@implementation SentryProfilingMutableState
+
+- (instancetype)init {
+    if (self = [super init]) {
+        _samples = [NSMutableArray<SentrySample *> array];
+        _stacks = [NSMutableArray<NSArray<NSNumber *> *> array];
+        _frames = [NSMutableArray<NSDictionary<NSString *, id> *> array];
+        _threadMetadata = [NSMutableDictionary<NSString *, NSMutableDictionary *> dictionary];
+        _queueMetadata = [NSMutableDictionary<NSString *, NSDictionary *> dictionary];
+        _frameIndexLookup = [NSMutableDictionary<NSString *, NSNumber *> dictionary];
+        _stackIndexLookup = [NSMutableDictionary<NSString *, NSNumber *> dictionary];
+    }
+    return self;
+}
+
+@end
+
+@implementation SentryProfilingState {
+    SentryProfilingMutableState *_mutableState;
+    std::mutex _lock;
+}
+
+- (instancetype)init
+{
+    if (self = [super init]) {
+        _mutableState = [[SentryProfilingMutableState alloc] init];
+    }
+    return self;
+}
+
+- (void)mutate:(void (^)(SentryProfilingMutableState *))block
+{
+    NSParameterAssert(block);
+    std::lock_guard<std::mutex> l(_lock);
+    block(_mutableState);
+}
+
+- (void)appendBacktrace:(const Backtrace &)backtrace {
+    [self mutate:^(SentryProfilingMutableState *state) {
+        const auto threadID = sentry_stringForUInt64(backtrace.threadMetadata.threadID);
```

Comment:
The logic inside `mutate` is essentially unchanged from the original logic of `processBacktrace` prior to the changes in #3018. The original code is here: https://github.com/getsentry/sentry-cocoa/blob/b7bd71993d0f05fd733eb36c665b7048a476e4fd/Sources/Sentry/SentryProfiler.mm#L91-L160

We now take a single lock to make the entire mutation which simplifies the code, there is no fine grained locking.